prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,208 @@
+				labelsString := ""
+				for _, label := range projectQuery.Query.Labels {
+					labelsString = fmt.Sprintf("%s label:%s", labelsString, label)
+				}
+				queryString := fmt.Sprintf("Query: \"org:%s state:%s%s\"", projectQuery.Query.Org, projectQuery.Query.State, labelsString)
+				configString = fmt.Sprintf("%s\n%s %s", configString, projectString, queryString)
+			}
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectQueries, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if len(projectQueries) == 0 {
+		return nil
+	}
+
+	// Only respond to label add or issue/PR open events
+	if !isPRChanged(pe) {
+		return nil
+	}
+
+	// Get any ProjectQueries that match this PR
+	matchedProjectQueries, err := getMatchingProjectQueries(gc, projectQueries, pe)
+	if err != nil {
+		return err
+	}
+	if len(matchedProjectQueries) == 0 {
+		return nil
+	}
+
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		le.WithError(err)
+	}
+
+	return nil
+}
+
+// labelsInLabels check that all of the queryLabels (the first list), are present in
+// the list of github.Labels (the second list)
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
+	for _, queryLabel := range queryLabels {
+		match := false
+		for _, label := range labels {
+			if queryLabel == label.Name {
+				match = true
+				break
+			}
+		}
+		if match == false {
+			return false
+		}
+	}
+	return true
+}
+
+// These are the only actions that this plugin cares about
+func isPRChanged(pe github.PullRequestEvent) bool {
+	switch pe.Action {
+	case github.PullRequestActionOpened:
+		return true
+	case github.PullRequestActionClosed:
+		return true
+	case github.PullRequestActionReopened:
+		return true
+	case github.PullRequestActionLabeled:
+		return true
+	case github.PullRequestActionUnlabeled:
+		return true
+	default:
+		return false
+	}
+}
+
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	matchedProjectQueries := make([]plugins.ProjectQuery, 0, 1)
+	// GetIssueLabels works for PRs as they are considered issues in the API
+	labels, err := gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
```

Comment:
This is one of the items I was asking about in the issue -- we need to be careful on how this consumes GitHub API tokens. I would suggest making this lazily resolve labels only when we are certain we need to know them to determine if a project should be added.

Reply:
I've changed this now to only use `GetIssueLabels` at the latest possible point, if required to determine a match.